app/src/main/java/org/jellyfin/mobile/utils/SystemUtils.kt

```diff
@@ -38,37 +42,57 @@ fun MainActivity.requestNoBatteryOptimizations() {
     }
 }
 
-fun MainActivity.requestDownload(uri: Uri, title: String, filename: String) {
-    val request = DownloadManager.Request(uri)
+suspend fun MainActivity.requestDownload(uri: Uri, title: String, filename: String) {
+    val requestWritePermissions = Build.VERSION.SDK_INT <= 28
+
+    if (requestWritePermissions) {
+        val granted = suspendCoroutine<Boolean> { continuation ->
+            requestPermission(WRITE_EXTERNAL_STORAGE) { requestPermissionsResult ->
+                continuation.resume(requestPermissionsResult[WRITE_EXTERNAL_STORAGE] == PERMISSION_GRANTED)
+            }
+        }
+
+        if (!granted) {
+            toast(R.string.download_no_storage_permission)
+            return
+        }
+    }
+
+    val downloadMethod = suspendCoroutine<Int> { continuation ->
+        val downloadMethod = appPreferences.downloadMethod
+
+        if (downloadMethod >= 0) continuation.resume(downloadMethod)
+        else runOnUiThread {
+            AlertDialog.Builder(this)
```

Comment:
What if a configuration change happens? (e.g. rotating the activity)
Maybe you could add a timeout to the coroutine to make sure it doesn't stay active forever? Or an `onDismissListener` maybe?